File: DSA With SUPREME 2.0 Batch/Array/RotateArray.cpp
// Example : I/P -> {2, 3, 5, 12, 65, 54}
// O/P -> {54, 2, 3, 5, 12, 65}

#include<iostream>
using namespace std;

void method1(int arr[],int size, int pos){
    int TempArr[pos];
    //Storing into Temp Array
    for(int i=size-pos ; i<size ; i++){
        TempArr[i- (size-pos)] = arr[i];
    }
    //shifting by kth position
    for(int i=size-1 ; i>pos-1 ; i--){
        arr[i] = arr[i-pos];
    }
    //storing element into original array from temp Array
    for(int i=0 ; i<pos ; i++){
        arr[i] = TempArr[i];
    }

    // Print an array
    for(int i=0 ; i<size ; i++){
        cout<<arr[i]<<" ";
    }
}

void method2(int arr[],int size, int pos){
     pos = pos % size;
    int TempArr[pos];

    // Storing into Temp Array
    for (int i = 0; i < size; i++) {
        int k = (i + pos) % size;
        TempArr[k] = arr[i];
    }

    int j = size - 1;

    // Copy elements from the end of arr to the beginning of TempArr
    for (int i = size - 1; i >= size - pos; i--) {
        TempArr[i] = arr[j--];
    }

    // Copy remaining elements from the beginning of arr to the end of TempArr
    for (int i = 0; i < size - pos; i++) {
        TempArr[i] = arr[i];
    }
}


int main(){
    int arr[] = {2, 3, 5, 12, 23, 32};
    // method1(arr, 6, 1);
    method2(arr, 6, 2);
    return 0;
}

File: DSA With SUPREME 2.0 Batch/01Array/RotateArray.java
/*
I/P ->{2, 3, 5, 12, 65, 54}, K=2
O/P -> {65, 54, 2, 3, 5, 12}
*/

class RotateArray{


    //brute force approach
    public static int[] rotateArray1(int arr[], int kth) {
        int size = arr.length;

        //Edge case
        if (kth < 0 || kth >= size) {
            throw new IllegalArgumentException("kth must be between 0 and size of array");
        } else {
            // Create a temporary array with size kth
            int[] temp = new int[kth];
    
            // Copy the last kth elements to temp
            for (int i = size - kth; i < size; i++) {
                temp[i - (size - kth)] = arr[i];
            }
    
            // Shift the remaining elements to the end
            for (int i = size - 1; i >= kth; i--) {
                arr[i] = arr[i - kth];
            }
    
            // Copy the temp elements back to the beginning
            for (int i = 0; i < kth; i++) {
                arr[i] = temp[i];
            }
    
            // Return the rotated array
            return arr;
        }
    }
    

    //Modulo approach
    public static int[] rotateArray2(int[] arr, int kth) {

        int size = arr.length;
        int[] temp = new int[size];
    
        // Edge case
        if (kth < 0 || kth >= size) {
            throw new IllegalArgumentException("kth must be between 0 and size of array");
        } else {

            for (int i = 0; i < size; i++) {

                // Calculate new index for rotation
                int newI = (i + kth) % size;

                // Place the element in the new index
                temp[newI] = arr[i];
            }


            // Return the rotated array
            return temp;
        }
    }
    
    
    public static void main(String[] args) {
        int[] arr = new int[]{2, 3, 5, 12, 23, 32};
        int[] shiftedArray = rotateArray2(arr, 2);
    
        for (int ele : shiftedArray) {
            System.out.print(ele + " ");
        }
    }
}  